energy-integration/governance-v2/src/lib.rs

```diff
@@ -0,0 +1,330 @@
+
+        proposal_id
+    }
+
+    /// Vote on a proposal. The voting power depends on the user's energy.
+    #[endpoint]
+    fn vote(&self, proposal_id: ProposalId) {
+        self.require_caller_not_self();
+        self.require_valid_proposal_id(proposal_id);
+        require!(
+            self.get_proposal_status(proposal_id) == GovernanceProposalStatus::Active,
+            "Proposal is not active"
+        );
+
+        let voter = self.blockchain().get_caller();
+        let new_user = self.user_voted_proposals(&voter).insert(proposal_id);
+        require!(new_user, ALREADY_VOTED_ERR_MSG);
+
+        let user_energy = self.get_energy_non_zero(voter.clone());
+        self.total_votes(proposal_id)
+            .update(|total_votes| *total_votes += &user_energy);
+
+        self.vote_cast_event(&voter, proposal_id, &user_energy);
+    }
+
+    /// Downvote a proposal. The voting power depends on the user's energy.
+    #[endpoint]
+    fn downvote(&self, proposal_id: ProposalId) {
+        self.require_caller_not_self();
+        self.require_valid_proposal_id(proposal_id);
+        require!(
+            self.get_proposal_status(proposal_id) == GovernanceProposalStatus::Active,
+            "Proposal is not active"
+        );
+
+        let downvoter = self.blockchain().get_caller();
+        let new_user = self.user_voted_proposals(&downvoter).insert(proposal_id);
+        require!(new_user, ALREADY_VOTED_ERR_MSG);
+
+        let user_energy = self.get_energy_non_zero(downvoter.clone());
+        self.total_downvotes(proposal_id)
+            .update(|total_downvotes| *total_downvotes += &user_energy);
+
+        self.downvote_cast_event(&downvoter, proposal_id, &user_energy);
+    }
+
+    /// Queue a proposal for execution.
+    /// This can be done only if the proposal has reached the quorum.
+    /// A proposal is considered successful and ready for queing if
+    /// total_votes - total_downvotes >= quorum
+    /// and all the required payments were deposited
+    #[endpoint]
+    fn queue(&self, proposal_id: ProposalId) {
+        self.require_caller_not_self();
+        require!(
+            self.get_proposal_status(proposal_id) == GovernanceProposalStatus::Succeeded,
+            "Can only queue succeeded proposals"
+        );
+        require!(
+            self.pending_payments_for_proposal(proposal_id).is_empty()
+                || !self.payments_depositor(proposal_id).is_empty(),
+            "Payments for proposal not deposited"
+        );
+
+        let current_block = self.blockchain().get_block_nonce();
+        self.proposal_queue_block(proposal_id).set(&current_block);
+
+        self.proposal_queued_event(proposal_id, current_block);
+    }
+
+    /// Execute a previously queued proposal.
+    /// This will clear the proposal and unlock the governance tokens.
+    /// Said tokens can then be withdrawn and used to vote/downvote other proposals.
```

Comment:
Again, I don't think this applies anymore, as we use energy to vote, instead of governance tokens.